File: tt.py
#tt.py

import chess

# Transposition Table Entry Types
EXACT = 0
LOWER_BOUND = 1  # Alpha bound (beta cutoff)
UPPER_BOUND = 2  # Beta bound (all moves failed low)

class TTEntry:
    __slots__ = ['zobrist', 'depth', 'score', 'flag', 'best_move']
    
    def __init__(self, zobrist, depth, score, flag, best_move):
        self.zobrist = zobrist
        self.depth = depth
        self.score = score
        self.flag = flag
        self.best_move = best_move

class TranspositionTable:
    def __init__(self, size_mb=256):
        """
        Initialize transposition table
        size_mb: Size in megabytes (default 256MB)
        """
        self._init_table(size_mb)
    
    def _init_table(self, size_mb):
        """Initialize or resize the table"""
        bytes_per_entry = 40
        self.size = (size_mb * 1024 * 1024) // bytes_per_entry
        self.table = [None] * self.size
        self.hits = 0
        self.misses = 0
    
    def resize(self, size_mb):
        """Resize the transposition table"""
        self._init_table(size_mb)
    
    def _get_hash(self, board):
        """Get zobrist hash - compatible with different python-chess versions"""
        if hasattr(board, 'zobrist_hash'):
            return board.zobrist_hash()
        elif hasattr(board, '_transposition_key'):
            return hash(board._transposition_key())
        else:
            return hash(board.fen())
    
    def index(self, zobrist_hash):
        """Get index for a zobrist hash"""
        return zobrist_hash % self.size
    
    def store(self, board, depth, score, flag, best_move=None):
        """Store with depth-preferred replacement"""
        zobrist = self._get_hash(board)
        idx = self.index(zobrist)
        
        existing = self.table[idx]
        
        # Replace if: empty, same position, or we searched deeper
        if existing is None or existing.zobrist == zobrist or existing.depth <= depth:
            self.table[idx] = TTEntry(zobrist, depth, score, flag, best_move)
    
    def probe(self, board, depth, alpha, beta):
        zobrist = self._get_hash(board)
        idx = self.index(zobrist)
        entry = self.table[idx]
        
        if entry is None or entry.zobrist != zobrist:
            self.misses += 1
            return False, 0, None
        
        # Found the position - always return the move for ordering
        best_move = entry.best_move
        
        # Can we use the score?
        if entry.depth >= depth:
            self.hits += 1
            
            if entry.flag == EXACT:
                return True, entry.score, best_move
            elif entry.flag == LOWER_BOUND and entry.score >= beta:
                return True, entry.score, best_move
            elif entry.flag == UPPER_BOUND and entry.score <= alpha:
                return True, entry.score, best_move
        
        # Can't use score, but return move for ordering
        self.misses += 1
        return False, 0, best_move

    def get_pv(self, board, max_depth=10):
        """Extract principal variation from transposition table"""
        pv = []
        seen_positions = set()
        
        for _ in range(max_depth):
            zobrist = self._get_hash(board)
            if zobrist in seen_positions:
                break
            seen_positions.add(zobrist)
            
            idx = self.index(zobrist)
            entry = self.table[idx]
            
            if entry is None or entry.zobrist != zobrist or entry.best_move is None:
                break
            
            if entry.best_move not in board.legal_moves:
                break
            
            pv.append(entry.best_move)
            board.push(entry.best_move)
        
        # Undo moves
        for _ in range(len(pv)):
            board.pop()
        
        return pv
    
    def clear(self):
        """Clear the transposition table"""
        self.table = [None] * self.size
        self.hits = 0
        self.misses = 0
    
    def get_stats(self):
        """Get hit/miss statistics"""
        total = self.hits + self.misses
        hit_rate = (self.hits / total * 100) if total > 0 else 0
        return {
            'hits': self.hits,
            'misses': self.misses,
            'hit_rate': hit_rate
        }

# Global transposition table
tt = TranspositionTable(size_mb=256)

File: src/tt.rs
use chess::{Board, ChessMove};

#[derive(Clone, Copy, PartialEq)]
pub enum TTFlag {
    Exact,
    LowerBound,
    UpperBound,
}

#[derive(Clone)]
pub struct TTEntry {
    pub zobrist: u64,
    pub depth: i32,
    pub score: i32,
    pub flag: TTFlag,
    pub best_move: Option<ChessMove>,
    pub age: u8,
}

pub struct TranspositionTable {
    pub table: Vec<Option<TTEntry>>,
    pub size: usize,
    pub mask: usize,
    pub hits: u64,
    pub misses: u64,
    pub generation: u8,
    pub filled_count: usize,
}

impl TranspositionTable {
    pub fn new(size_mb: usize) -> Self {
        let bytes_per_entry = std::mem::size_of::<Option<TTEntry>>();
        let entries = (size_mb * 1024 * 1024) / bytes_per_entry;
        
        let size = if entries.is_power_of_two() {
            entries
        } else {
            entries.next_power_of_two() / 2
        };
        
        Self {
            table: vec![None; size],
            size,
            mask: size - 1,
            hits: 0,
            misses: 0,
            generation: 0,
            filled_count: 0,
        }
    }

    #[inline]
    fn index(&self, zobrist: u64) -> usize {
        (zobrist as usize) & self.mask
    }

    pub fn new_search(&mut self) {
        self.generation = self.generation.wrapping_add(1);
    }

    #[inline]
    fn is_over_capacity(&self) -> bool {
        (self.filled_count * 100) / self.size > 95
    }

    pub fn store(&mut self, board: &Board, depth: i32, score: i32, flag: TTFlag, best_move: Option<ChessMove>, ply: i32) {
        let zobrist = board.get_hash();
        let idx = self.index(zobrist);
        
        let adjusted_score = if score > 90000 {
            score + ply
        } else if score < -90000 {
            score - ply
        } else {
            score
        };
        
        let is_over_capacity = self.is_over_capacity();
        
        let should_replace = match &self.table[idx] {
            None => true,
            Some(entry) => {
                if entry.zobrist == zobrist {
                    true
                } else {
                    let age_diff = self.generation.wrapping_sub(entry.age);
                    
                    if is_over_capacity {
                        age_diff >= 5 || depth > entry.depth
                    } else {
                        age_diff >= 10 || depth >= entry.depth
                    }
                }
            }
        };

        if should_replace {
            let was_empty = self.table[idx].is_none();
            
            self.table[idx] = Some(TTEntry {
                zobrist,
                depth,
                score: adjusted_score,
                flag,
                best_move,
                age: self.generation,
            });
            
            if was_empty {
                self.filled_count += 1;
            }
        }
    }

    pub fn probe(&mut self, board: &Board, depth: i32, alpha: i32, beta: i32, ply: i32) -> (bool, i32, Option<ChessMove>) {
        let zobrist = board.get_hash();
        let idx = self.index(zobrist);

        match &self.table[idx] {
            None => {
                self.misses += 1;
                (false, 0, None)
            }
            Some(entry) => {
                if entry.zobrist != zobrist {
                    self.misses += 1;
                    return (false, 0, None);
                }

                let best_move = entry.best_move;

                if entry.depth >= depth {
                    self.hits += 1;
                    
                    let adjusted_score = if entry.score > 90000 {
                        entry.score - ply
                    } else if entry.score < -90000 {
                        entry.score + ply
                    } else {
                        entry.score
                    };
                    
                    let use_score = match entry.flag {
                        TTFlag::Exact => true,
                        TTFlag::LowerBound => adjusted_score >= beta,
                        TTFlag::UpperBound => adjusted_score <= alpha,
                    };

                    if use_score {
                        return (true, adjusted_score, best_move);
                    }
                }

                self.misses += 1;
                (false, 0, best_move)
            }
        }
    }

    pub fn clear(&mut self) {
        self.table = vec![None; self.size];
        self.hits = 0;
        self.misses = 0;
        self.filled_count = 0;
    }

    pub fn get_stats(&self) -> (u64, u64, f64) {
        let total = self.hits + self.misses;
        let hit_rate = if total > 0 {
            (self.hits as f64 / total as f64) * 100.0
        } else {
            0.0
        };
        (self.hits, self.misses, hit_rate)
    }
}